src/acceleration/Acceleration.cpp

```diff
@@ -17,16 +17,22 @@ void Acceleration::checkDataIDs(const DataMap &cplData) const
 void Acceleration::applyRelaxation(double omega, const DataMap &cplData) const
 {
   for (const DataMap::value_type &pair : cplData) {
-    const auto  couplingData = pair.second;
-    auto &      values       = couplingData->values();
-    const auto &oldValues    = couplingData->previousIteration();
-    values *= omega;
-    values += oldValues * (1 - omega);
-    if (couplingData->hasGradient()) {
-      auto &      gradients    = couplingData->gradients();
-      const auto &oldGradients = couplingData->previousIterationGradients();
-      gradients *= omega;
-      gradients += oldGradients * (1 - omega);
+    const auto couplingData = pair.second;
+
+    for (auto &stample : couplingData->stamples()) {
+      auto values            = stample.sample.values;
+      auto oldValues         = couplingData->getPreviousValuesAtTime(stample.timestamp); // IMPORTANT DETAIL: The interpolation that we use for resampling does not necessarily have to be the same interpolation as the interpolation the user accesses via read-data. (But probably it is easier to just use the same)
+      couplingData->values() = values * omega;
+      couplingData->values() += oldValues * (1 - omega);
+
+      if (couplingData->hasGradient()) {
+        auto       gradients      = stample.sample.gradients;
+        const auto oldGradients   = couplingData->getPreviousGradientsAtTime(stample.timestamp); // IMPORTANT DETAIL: The interpolation that we use for resampling does not necessarily have to be the same interpolation as the interpolation the user accesses via read-data. (But probably it is easier to just use the same)
+        couplingData->gradients() = gradients * omega;
+        couplingData->gradients() += oldGradients * (1 - omega);
+      }
+
+      couplingData->setSampleAtTime(stample.timestamp, couplingData->sample());
```

Comment:
This now also happens if `substeps==false`, right? Or how is this case treated?

Reply:
Yes. But if `substeps==false`, the functions still return reasonable values (if `substeps==false`, just a single value at `WINDOW_END`).

Reply:
Because if `substeps==false`, we don't even collect anything in `stamples()`, correct?